### DISCUSSION TITLE
Why deserialize json to decimal suffer from floating point precision error?

### DISCUSSION THREAD

**Discussion by sn4k3** [dotnet/runtime#90428] [Q&A]
I'm deserializing json to class, and using decimal type to get the exact number as set on json, but the number is set like the float precision error. 

Runtime: NET 7.0 

```C#
public class GazetteerRecord
{
    [JsonPropertyName("MRGID")]
    public int MrgId { get; set; }

    [JsonPropertyName("gazetteerSource")]
    public string GazetteerSource { get; set; } = string.Empty;

    [JsonPropertyName("placeType")]
    public string PlaceType { get; set; } = string.Empty;

    [JsonPropertyName("latitude")]
    public decimal? Latitude { get; set; } = null;

    [JsonPropertyName("longitude")]
    public decimal? Longitude { get; set; } = null;

    [JsonPropertyName("minLatitude")]
    public decimal? MinLatitude { get; set; } = null;

    [JsonPropertyName("minLongitude")]
    public decimal? MinLongitude { get; set; } = null;

    [JsonPropertyName("maxLatitude")]
    public decimal? MaxLatitude { get; set; } = null;

    [JsonPropertyName("maxLongitude")]
    public decimal? MaxLongitude { get; set; } = null;

    [JsonPropertyName("precision")]
    public decimal? Precision { get; set; } = null;

    [JsonPropertyName("preferredGazetteerName")]
    public string PreferredGazetteerName { get; set; } = string.Empty;

    [JsonPropertyName("preferredGazetteerNameLang")]
    public string PreferredGazetteerNameLang { get; set; } = string.Empty;

    [JsonPropertyName("status")]
    public string Status { get; set; } = string.Empty;

    [JsonPropertyName("accepted")]
    public int Accepted { get; set; }


    /// <inheritdoc />
    public override string ToString()
    {
        return $"{nameof(MrgId)}: {MrgId}, {nameof(GazetteerSource)}: {GazetteerSource}, {nameof(PlaceType)}: {PlaceType}, {nameof(Latitude)}: {Latitude}, {nameof(Longitude)}: {Longitude}, {nameof(MinLatitude)}: {MinLatitude}, {nameof(MinLongitude)}: {MinLongitude}, {nameof(MaxLatitude)}: {MaxLatitude}, {nameof(MaxLongitude)}: {MaxLongitude}, {nameof(Precision)}: {Precision}, {nameof(PreferredGazetteerName)}: {PreferredGazetteerName}, {nameof(PreferredGazetteerNameLang)}: {PreferredGazetteerNameLang}, {nameof(Status)}: {Status}, {nameof(Accepted)}: {Accepted}";
    }
}
```

 ```json
{
    "MRGID": 2554,
    "gazetteerSource": "ASFA thesaurus",
    "placeType": "Coast",
    "latitude": 51.15,
    "longitude": 2.7,
    "minLatitude": null,
    "minLongitude": null,
    "maxLatitude": null,
    "maxLongitude": null,
    "precision": null,
    "preferredGazetteerName": "Belgian West Coast",
    "preferredGazetteerNameLang": "English",
    "status": "standard",
    "accepted": 2554
  }
```

Print result:

```
MrgId: 2554
GazetteerSource: ASFA thesaurus
PlaceType: Coast
Latitude: 51,149999999999999
Longitude: 2,7000000000000002
MinLatitude:
MinLongitude:
MaxLatitude:
MaxLongitude:
Precision:
PreferredGazetteerName: Belgian West Coast
PreferredGazetteerNameLang: English
Status: standard
Accepted: 2554
```

![image](https://github.com/dotnet/runtime/assets/113281/72af6915-142a-4d64-8467-b7bd48afc2c4)

Notice the `Latitude` and `Longitude` with the precision error instead of same values from json.
How can I improve this situation?

**Comment by GrabYourPitchforks** [comment]
/cc @dotnet/area-system-text-json for thoughts. I could not repro this in sharplab; the app outputs `Latitude: 51.15, Longitude: 2.7` as expected.

\[[sharplab link](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEUCuA7AHwAEAmABgFgAoUgRmqLIAIjaA6AFRgQzYCkAzhDwBuBs1aduvQcLYBlGFACWAQwA2ygF6qMy4WKrV5GFXgDmTAUwC8TAESPqAb2pN39gLIAlAOIBJABF7ECYSAFZw9DcPc1UtGAwMGCV5CBwoMBgQhwBBeQAxXKYMAAsYAVUMnAF7NBiHAAd1VSyOAE9G7ND7AGEIVQEMOob7Fr0MHAATbqZw9lpw+qp3B3Vhc2VJmZySNgB2ZdX7AFtlPAAZXS3p2bwcdXUjjzPLjZud0PvH54cT1QQVwmtxy3yeo3+gPe2zuD3BKw8jVgYGUAn0eFBcN+9iRMAAZkpYFNfPFEsklAA5VQnWb2ABCMHUm1UeCYAHUKhgmP1BsNsbiCVAiSSEkkUlAqTSrhYcvYAKIWTQCUojBEOIa6GqyjV4KaqKBTVXHVpZRrJQ2hCJRNwAX2ojnshlWTGoADd9UwIMAAFa2JiyPCKFQabRKNiBCpKNSaBIAHhFZPF3hgkANAD4ABQCACUhlYAE4M17vbnqAwAMwsEhMBNipTJ1NTFwNJgAbQDAAUoBAulAMO1JTAM14/EF7NmALotoiV85czxQcz+KZMZxMcyJERWTdMO1GNVtzvd3v9wfDuKi8lQNIZLLjqcHmcsWjMWtXm+ZGCr9c7gQ7m1+qwZBsHKJxmu0hgtu2Qh4F2PZKKe1JDjiLRtJ02STtOlZAUwHaoTAHRdN+G4YFuf6kbugEviBYH9pBB7QcIcEngOSHDuMHwYQ+zpPjMKL/OoAD8TBApxxG/v+fpgvRzqMbBx4IaxNLsdCIKYY+lZ8coAnCRcqkzOJFHkVuAF2NJZYMUe8F9kpyGvKJML3lhTBaTpTCeOcDm3IZZGSWZcIyascnMYpZ6nJ5+lcc5rkaMJHlvBYYlriRvkUaZTDmfuslWSxYWQl5OzqTxmkptpsXuQCBVfslElpVJAUWdlMEhTZeWVZFTkaS5pVuZ47WJTCPnbnV/mPIF7jBQprVsTiyKouinXFd1/HlV2pVosIQ3GZRo3qONh7NVNiHKbN+KEjAxKknWEpIYtqxPjha2CsKV1XoOW1+c+wGgeB+2TdZx3IQK52XZe4qDtK5h3e4D0vrhsDPRdb7g0hkMfSNX00b9jVBTloUzRqky1EV93YXDJiatYNVGZ9QFY3ROMTXj00nSaMBmhd0MsLOeBcrkYCmua6MmRZLYAPQS0wsbnOUKgYFMEBgEwYtps5ECuoSygGThHAQCYZjmBm2Ytq4B48fsTAACT2M4eBIRAeIZguS5TNmNqhM4zvLjaaCrnbNIOxmyOpOkn5ux7wfXqHWQ+379uO3hrQEeh4eronaFdLHtvxxmVWp84VVZ/7MCB3pA23PnZebDCRc5/Fefu6u9fXDXvvZwHjv15F+dd+XMy1x3TuVS3FeN57w/Av3bfF4HfVQn3MA9/11e3APJcJ3NG14Pna0olva+B09IOR4OO8I8fr0ozSB8b2dQpI5flKoyy5hn3fL1g0/UovzfGYU0T+d/41F/vzQWF186gPZuaG0joWx7htEAA=)\]

**Comment by krwq** [comment]
JSON defers to Utf8Parser:
https://github.com/dotnet/runtime/blob/main/src/libraries/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.TryGet.cs#L1211

which ends up calling:
https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Utf8Parser/Utf8Parser.Decimal.cs

parsing logic being here: https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/Number.Parsing.cs#L983

@tannergooding might now perhaps - that logic is somewhat... optimized 😄

To be perfectly honest though I'm not seeing the behavior you're seeing:

<details>
<summary>Click to see code</summary>

```csharp
using System.Text.Json;
using System.Text.Json.Serialization;

string json = """
    {
      "MRGID": 2554,
      "gazetteerSource": "ASFA thesaurus",
      "placeType": "Coast",
      "latitude": 51.15,
      "longitude": 2.7,
      "minLatitude": null,
      "minLongitude": null,
      "maxLatitude": null,
      "maxLongitude": null,
      "precision": null,
      "preferredGazetteerName": "Belgian West Coast",
      "preferredGazetteerNameLang": "English",
      "status": "standard",
      "accepted": 2554
    }
    """;

GazetteerRecord record = JsonSerializer.Deserialize<GazetteerRecord>(json);
Console.WriteLine(record);

public class GazetteerRecord
{
    [JsonPropertyName("MRGID")]
    public int MrgId { get; set; }

    [JsonPropertyName("gazetteerSource")]
    public string GazetteerSource { get; set; } = string.Empty;

    [JsonPropertyName("placeType")]
    public string PlaceType { get; set; } = string.Empty;

    [JsonPropertyName("latitude")]
    public decimal? Latitude { get; set; } = null;

    [JsonPropertyName("longitude")]
    public decimal? Longitude { get; set; } = null;

    [JsonPropertyName("minLatitude")]
    public decimal? MinLatitude { get; set; } = null;

    [JsonPropertyName("minLongitude")]
    public decimal? MinLongitude { get; set; } = null;

    [JsonPropertyName("maxLatitude")]
    public decimal? MaxLatitude { get; set; } = null;

    [JsonPropertyName("maxLongitude")]
    public decimal? MaxLongitude { get; set; } = null;

    [JsonPropertyName("precision")]
    public decimal? Precision { get; set; } = null;

    [JsonPropertyName("preferredGazetteerName")]
    public string PreferredGazetteerName { get; set; } = string.Empty;

    [JsonPropertyName("preferredGazetteerNameLang")]
    public string PreferredGazetteerNameLang { get; set; } = string.Empty;

    [JsonPropertyName("status")]
    public string Status { get; set; } = string.Empty;

    [JsonPropertyName("accepted")]
    public int Accepted { get; set; }


    /// <inheritdoc />
    public override string ToString()
    {
        return $"{nameof(MrgId)}: {MrgId}, {nameof(GazetteerSource)}: {GazetteerSource}, {nameof(PlaceType)}: {PlaceType}, {nameof(Latitude)}: {Latitude}, {nameof(Longitude)}: {Longitude}, {nameof(MinLatitude)}: {MinLatitude}, {nameof(MinLongitude)}: {MinLongitude}, {nameof(MaxLatitude)}: {MaxLatitude}, {nameof(MaxLongitude)}: {MaxLongitude}, {nameof(Precision)}: {Precision}, {nameof(PreferredGazetteerName)}: {PreferredGazetteerName}, {nameof(PreferredGazetteerNameLang)}: {PreferredGazetteerNameLang}, {nameof(Status)}: {Status}, {nameof(Accepted)}: {Accepted}";
    }
}
```

</details>

The output I'm seeing is this:

```
MrgId: 2554, GazetteerSource: ASFA thesaurus, PlaceType: Coast, Latitude: 51.15, Longitude: 2.7, MinLatitude: , MinLongitude: , MaxLatitude: , MaxLongitude: , Precision: , PreferredGazetteerName: Belgian West Coast, PreferredGazetteerNameLang: English, Status: standard, Accepted: 2554
```

Can you share environment you're running this on? (i.e. `dotnet --info` output) and how you're building/running app along with your source code?

Also note you can always write JsonConverter when you're not happy with specific serialization logic for a type although in this case it's rather strange you're seeing different output

**Comment by sn4k3** [comment]
For the reference the project can be found here [SpeciesDatabaseApi](https://github.com/sn4k3/SpeciesDatabaseApi)

After some investigation I found that most probably is my mistake, the data from [JSON rest service](https://www.marineregions.org/rest/getGazetteerRecordByMRGID.json/2554/) already comes with that error on precision (at least for C# request), and C# lib is correctly assigning the value to the model. I saw this behavior in multiple and different calls, but I've compared the data from browser display rather than what I actual got from C# request.
Writing the JSON to be deserialised to the terminal prove that data is indeed that...

`.\SpeciesDatabaseCmd.exe MARINEREGIONS Record 2554`
![image](https://github.com/dotnet/runtime/assets/113281/a25782bd-8eb7-4c6d-b623-d94f1d055906)

But if we open the request URL on browser I see stable numbers instead

![image](https://github.com/dotnet/runtime/assets/113281/db771760-c157-4be9-9f67-0e03b5eaf3bb)

Now which one is lying about data? Edge browser or C#?

After, I sent same request using curl and WSL, it results in same data as C# gets
![image](https://github.com/dotnet/runtime/assets/113281/3d58e7f3-cf86-4b63-a790-69d2a24d38f1)

This makes me think that Edge somehow round and make JSON prettier resulting in my confusion about the data as I used it to compare the results I got into the model...
In the end nothing wrong with C# or it libraries, and a lesson learnt (Never use browser display to compare data ^^)

my dotnet info:

```
.NET SDK:
 Version:   7.0.400
 Commit:    73bf45718d

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.22621
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\7.0.400\

Host:
  Version:      7.0.10
  Architecture: x64
  Commit:       a6dbb800a4

.NET SDKs installed:
  7.0.400 [C:\Program Files\dotnet\sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.App 3.1.32 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.21 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 7.0.10 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 3.1.32 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.21 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 7.0.10 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 6.0.21 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 7.0.10 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]

Other architectures found:
  x86   [C:\Program Files (x86)\dotnet]
    registered at [HKLM\SOFTWARE\dotnet\Setup\InstalledVersions\x86\InstallLocation]

Environment variables:
  Not set

global.json file:
  Not found

Learn more:
  https://aka.ms/dotnet/info

Download .NET:
  https://aka.ms/dotnet/download
```
